File: frontend/src/types/problemDetails.ts
/**
 * RFC 7807 Problem Details フロントエンド型定義
 */
export interface ProblemDetails {
  type: string;
  title: string;
  status: number;
  detail?: string;
  instance?: string;
  invalid_params?: Array<{
    name: string;
    reason: string;
  }>;
}

export function isProblemDetails(error: unknown): error is ProblemDetails {
  return (
    typeof error === "object" &&
    error !== null &&
    "title" in error &&
    "status" in error
  );
}

File: src/backend/schemas/problem_details.py
"""RFC 7807 Problem Details スキーマ。"""
from __future__ import annotations
from typing import Any, Optional
from pydantic import BaseModel, Field


class ProblemDetails(BaseModel):
    """RFC 7807 に準拠した構造化エラーレスポンス。"""
    type: str = Field(
        default="about:blank",
        description="エラー種別を識別するURI参照",
    )
    title: str = Field(
        ...,
        description="人間が読める簡潔なエラー要約",
    )
    status: int = Field(
        ...,
        description="HTTPステータスコード",
    )
    detail: Optional[str] = Field(
        default=None,
        description="この発生インスタンスに特有の人道的な詳細説明",
    )
    instance: Optional[str] = Field(
        default=None,
        description="エラーが発生した具体的なリソースまたはリクエストパス",
    )
    invalid_params: Optional[list[dict[str, Any]]] = Field(
        default=None,
        description="バリデーションエラー時のフィールド別詳細",
    )
